File: Design Patterns/Creational/Prototype/prototype_test.go
package prototype

import "testing"

func TestUserClone(t *testing.T) {
	original := &User{
		UserID:      1,
		Username:    "johndoe",
		Email:       "john@example.com",
		DisplayName: "John Doe",
		Age:         25,
	}

	cloned := original.Clone().(*User)

	// Verify all fields are copied
	if cloned.UserID != original.UserID ||
		cloned.Username != original.Username ||
		cloned.Email != original.Email ||
		cloned.DisplayName != original.DisplayName ||
		cloned.Age != original.Age {
		t.Error("cloned user should have the same attribute values as the original")
	}

	// Verify it's a different object in memory
	if original == cloned {
		t.Error("cloned user should not be the same pointer as the original")
	}
}

func TestUserCloneIsIndependent(t *testing.T) {
	original := &User{UserID: 1, Username: "alice", Email: "alice@example.com", DisplayName: "Alice", Age: 30}
	cloned := original.Clone().(*User)

	// Mutating the clone should not affect the original
	cloned.Username = "bob"
	cloned.Age = 40

	if original.Username == cloned.Username {
		t.Error("modifying the clone should not affect the original")
	}
	if original.Age == cloned.Age {
		t.Error("modifying the clone's age should not affect the original")
	}
}

func TestRegistryAddAndClone(t *testing.T) {
	registry := NewUserPrototypeRegistry()

	admin := &User{UserID: 100, Username: "admin", Email: "admin@platform.com", DisplayName: "Admin User", Age: 35}
	member := &User{UserID: 200, Username: "member", Email: "member@platform.com", DisplayName: "Regular Member", Age: 22}

	registry.AddPrototype("admin", admin)
	registry.AddPrototype("member", member)

	// Clone an admin user
	clonedAdmin := registry.Clone("admin")
	if clonedAdmin == nil {
		t.Fatal("cloned admin should not be nil")
	}
	if clonedAdmin == admin {
		t.Error("cloned admin should be a different object than the prototype")
	}
	if clonedAdmin.Username != "admin" || clonedAdmin.Age != 35 {
		t.Error("cloned admin should have the same attributes as the prototype")
	}

	// Clone a member user
	clonedMember := registry.Clone("member")
	if clonedMember == nil {
		t.Fatal("cloned member should not be nil")
	}
	if clonedMember.Email != "member@platform.com" {
		t.Error("cloned member should have the same email as the prototype")
	}
}

func TestRegistryGetPrototype(t *testing.T) {
	registry := NewUserPrototypeRegistry()

	user := &User{UserID: 1, Username: "test", Email: "test@example.com", DisplayName: "Test", Age: 20}
	registry.AddPrototype("test", user)

	retrieved := registry.GetPrototype("test")
	if retrieved != user {
		t.Error("GetPrototype should return the original prototype pointer")
	}
}

func TestRegistryCloneNonExistent(t *testing.T) {
	registry := NewUserPrototypeRegistry()

	result := registry.Clone("nonexistent")
	if result != nil {
		t.Error("cloning a non-existent type should return nil")
	}
}

func TestRegistryCloneIndependence(t *testing.T) {
	registry := NewUserPrototypeRegistry()
	registry.AddPrototype("default", &User{UserID: 1, Username: "default", Email: "default@example.com", DisplayName: "Default", Age: 25})

	clone1 := registry.Clone("default")
	clone2 := registry.Clone("default")

	// Mutate clone1
	clone1.Username = "modified"
	clone1.Age = 99

	// clone2 and the original prototype should be unaffected
	if clone2.Username != "default" || clone2.Age != 25 {
		t.Error("clones should be independent of each other")
	}

	original := registry.GetPrototype("default")
	if original.Username != "default" || original.Age != 25 {
		t.Error("cloning should not modify the original prototype")
	}
}

File: Projects/outOfStockNotification/Design Patterns/Creational/Prototype/PrototypeTest.java
package prototype;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class PrototypeTest {

    @Test
    void testUserClone() {
        User original = new User(1, "johndoe", "john@example.com", "John Doe", 25);
        User cloned = original.clone();

        assertEquals(original.getUserId(), cloned.getUserId());
        assertEquals(original.getUsername(), cloned.getUsername());
        assertEquals(original.getEmail(), cloned.getEmail());
        assertEquals(original.getDisplayName(), cloned.getDisplayName());
        assertEquals(original.getAge(), cloned.getAge());
        assertNotSame(original, cloned);
    }

    @Test
    void testUserCloneIsIndependent() {
        User original = new User(1, "alice", "alice@example.com", "Alice", 30);
        User cloned = original.clone();

        cloned.setUsername("bob");
        cloned.setAge(40);

        assertNotEquals(original.getUsername(), cloned.getUsername());
        assertNotEquals(original.getAge(), cloned.getAge());
    }

    @Test
    void testRegistryAddAndClone() {
        UserPrototypeRegistry registry = new UserPrototypeRegistryImpl();

        User admin = new User(100, "admin", "admin@platform.com", "Admin User", 35);
        User member = new User(200, "member", "member@platform.com", "Regular Member", 22);

        registry.addPrototype("admin", admin);
        registry.addPrototype("member", member);

        User clonedAdmin = registry.clone("admin");
        assertNotNull(clonedAdmin);
        assertNotSame(admin, clonedAdmin);
        assertEquals("admin", clonedAdmin.getUsername());
        assertEquals(35, clonedAdmin.getAge());

        User clonedMember = registry.clone("member");
        assertNotNull(clonedMember);
        assertEquals("member@platform.com", clonedMember.getEmail());
    }

    @Test
    void testRegistryGetPrototype() {
        UserPrototypeRegistry registry = new UserPrototypeRegistryImpl();
        User user = new User(1, "test", "test@example.com", "Test", 20);
        registry.addPrototype("test", user);

        assertSame(user, registry.getPrototype("test"));
    }

    @Test
    void testRegistryCloneNonExistent() {
        UserPrototypeRegistry registry = new UserPrototypeRegistryImpl();
        assertNull(registry.clone("nonexistent"));
    }

    @Test
    void testRegistryCloneIndependence() {
        UserPrototypeRegistry registry = new UserPrototypeRegistryImpl();
        registry.addPrototype("default", new User(1, "default", "default@example.com", "Default", 25));

        User clone1 = registry.clone("default");
        User clone2 = registry.clone("default");

        clone1.setUsername("modified");
        clone1.setAge(99);

        assertEquals("default", clone2.getUsername());
        assertEquals(25, clone2.getAge());

        User original = registry.getPrototype("default");
        assertEquals("default", original.getUsername());
        assertEquals(25, original.getAge());
    }
}
